### ISSUE TITLE
Add: shipping boxes "fit" algorithm for shipping modules to use

### ISSUE THREAD

**Issue by drbyte** [zencart/zencart#2934]
Most shipping services can only quote on number of boxes based on size/weight/destination.
They usually cannot calculate how to "best fit all the cart items" into the best combination of box sizes available to the storeowner/packager.

ZC core currently just fits items based on max-weight-per-package. This isn't ideal as it's not thorough. It also doesn't account for packages that will "ship in their own box" instead of being combined into other boxes with other items.

**Comment by g2ktcf** [comment]
I have some thoughts on this and some code started based on Largest Area Fit First method (LAFF).  I have seen some C# implementations of LAFF already although I think that they have some holes to be considered.  All the bin packing solutions I have seen and reviewed all admit that "human packers can do better" so I think it would be wise to do this in a two path process.  1) Retrieve "memorized shipments" based on experience and if this cannot be located then 2) use bin packing to get as close as possible.

My real question is I think there needs to be a notification to trigger start the "packing" process.  This would not be super fast as the number of calculations can be significant.   Then, the packing process would return a shipment object and/or JSON with the details.    I am unsure of how other shipping modules would incorporate this however.

Just wondering if anyone has any thoughts on this.

**Comment by torvista** [comment]
A comprehensive solution is indeed complex and the time and effort required to get it right v. the benefits really depend on your business and if the work involved in coding this compensates for the lost money on some shipping costs for some items occasionally. We just accept those few euros sometimes as not significant in the grand scheme of costs.

I use Advanced Shipper anyway, which allows specific products to be exceptions to the standard rules.
Any solution needs exceptions for bulky items where the carrier will charge for the volumetric weight.

**Comment by lat9** [comment]
That notification could be at the end of the checkout_shipping: `NOTIFY_HEADER_END_CHECKOUT_SHIPPING`.

At that point, the order's been built (albeit without payment information) so that an `$order` object is available to gather the products themselves and the base processing has determined that the checkout can continue.

**Comment by g2ktcf** [comment]
> A comprehensive solution is indeed complex and the time and effort required to get it right v. the benefits really depend on your business and if the work involved in coding this compensates for the lost money on some shipping costs for some items occasionally. We just accept those few euros sometimes as not significant in the grand scheme of costs.
> 
> I use Advanced Shipper anyway, which allows specific products to be exceptions to the standard rules. Any solution needs exceptions for bulky items where the carrier will charge for the volumetric weight.

Agreed, what I have planned will do all of that IF the shipping modules take advantage of the information this provides.

**Comment by g2ktcf** [comment]
> That notification could be at the end of the checkout_shipping: `NOTIFY_HEADER_END_CHECKOUT_SHIPPING`.
> 
> At that point, the order's been built (albeit without payment information) so that an `$order` object is available to gather the products themselves and the base processing has determined that the checkout can continue.

has any "shipping" method been presented to the customer at that point?  The hope for me is this is JUST before presenting the shipping options to the customer.

**Comment by lat9** [comment]
> > That notification could be at the end of the checkout_shipping: `NOTIFY_HEADER_END_CHECKOUT_SHIPPING`.
> > At that point, the order's been built (albeit without payment information) so that an `$order` object is available to gather the products themselves and the base processing has determined that the checkout can continue.
> 
> has any "shipping" method been presented to the customer at that point? The hope for me is this is JUST before presenting the shipping options to the customer.
Nope, that notification comes after the header's completed processing; no rendering has yet been made.

**Comment by g2ktcf** [comment]
The shipping modules would need to be updated for this to work.  As I understand it right now, those modules simply pass a weight and a shipment type.  This is just "what I am thinking" and any input is appreciated.

The plan here would be to:

1. Have a list of MEMORIZED shipment items.  Everything I have read said humans can do a better job.  So for repeated shipments, why not "just look them up".  Query the dbase and create the $shipment object for the shipping modules off the returned data.  This would be very quick.  There can even be multiple options and we can compare an show the most or least expensive.  However, this depends on how the shipping modules would interface.  Any shipping module writer's input would be great.  I could see this module incorporating those modules...but that is for another day. 
2. Have the store owner maintain a list of shipping box types and sizes.  Most owners use a specific set of boxes and this is the first cut to control how many calculations need to be done.
2. The product table would be modified to include item dimensions in rectangular form.  I am not planning on trying to incorporate anything other than rectangle shaped objects at this time.  Placing an irregularly shaped object is not hugely difficult but finding out how to use the "interstitial" spaces would be a bear.
3. Pull the order items from $orders  and get the dimensions from the product table.
4. Query the db for all boxes with a volume large enough and a longest side that matches order items.
5. ...magic happens.... (this is the LAFF, heuristics based algorithm)
6. create the $shipments object that contains all shipments that can be used.  
7. Set a notifier to let ZC know that the process is completed.
8. Return $shipments object.  If no results were found either memorized or calculated, then $shipments = NULL;


What I have done:

I am working in isolation, outside of ZC for now.  

1. I have created an "insertion points" object that contains possible locations (as 3D points)  for new items to be placed and their volumes.
2. I have created a general 3d box object for that is the base for the shipping box, and the base for the items.
3. I have a 3D collision code based off the 2D AABB process.  This is method is defined in a new class called $items that extends the 3D box object.  I am currently trying to create a full fledged testing process to fully validate this process.  I have done some testing but I am not sure I have been able to identify the outliers on it.
4. I have an outline of the LAFF process.  I am not 100% convinced it gets really good results as it does not rotate boxes to see if they are a better fit.  So, I am using the base paper to create my own version.
5. I expect to get the entire thing working first.  Once that is done, I will need some assistance/advice on how to integrate fully with ZC and other shipping modules.
